text/3389-manifest-lint.md

```diff
@@ -0,0 +1,250 @@
+- We have not yet defined semantics for sharing something like this across a
+  workspace
+
+Instead of using workspace inheritance for `[lint]`, we could make it
+workspace-level configuration, like `[patch]` which is automatically applied to
+all workspace members.  However, `[patch]` and friends are because they affect
+the resolver / `Cargo.toml` and so they can only operate at the workspace
+level.  `[lints]` is more like `[dependencies]` in being something that applies
+at the package level but we want shared across workspaces.
+
+Instead of traditional workspace inheritance where there is a single value to
+inherit with `workspace = true`, we could have `[workspace.lints.<preset>]`
+which defines presets and the user could do `lints.<preset> = true`.  The user
+could then name them as they wish to avoid collision with rustc lints.
+
+Instead of the `[lints]` table being `lint = "level"`, we could organize
+it around `level = ["lint", ...]` like some other linters do (like
+[ruff](https://beta.ruff.rs/docs/configuration/)) but this works better for
+logically organizing lints, highlighting what changed in diffs, and for
+possibly adding lint-specific configuration in the future.
+
+# Prior art
+[prior-art]: #prior-art
+
+Python
+- [flake8](https://flake8.pycqa.org/en/latest/user/configuration.html)
+  - Format is `level = [lint, ...]`
+- [pylint](https://github.com/PyCQA/pylint/blob/main/examples/pylintrc#L402)
+  - Format is `level = [lint, ...]` but the [config file is a reflection of the CLI](https://pylint.pycqa.org/en/latest/user_guide/configuration/index.html)
+- [ruff](https://beta.ruff.rs/docs/configuration/)
+  - Format is `level = [lint, ...]`, likely due to past precedence in ecosystem (see above)
+
+Javascript
+- [eslint](https://eslint.org/docs/latest/use/configure/rules)
+  - Format is `lint = level` / `lint = [ level, additional config ]`
+
+Go
+- [golangci-lint](https://golangci-lint.run/usage/configuration/)
+  - Format is `level = [lint, ...]`
+
+Ruby
+- [rubocop](https://docs.rubocop.org/rubocop/1.45/configuration.html)
+  - Format is `Lint: Enabled: true`
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+How should we hand rustdoc lint levels or, in the future, cargo lint levels?
+The current proposal takes all lints and passes them to rustc like `RUSTFLAGS`
+but rustdoc uses `RUSTDOCFLAGS` and cargo would use neither.  This also starts
+to get into
+[user-defined tool attributes](https://rust-lang.github.io/rfcs/2103-tool-attributes.html).
+
+Should we only apply/fingerprint lints for the appropriate tool?  For example,
+we would not include and fingerprint `clippy::` lints when running builds,
+allowing them to change without forcing a rebuild.  We likely already need to
+be tool-aware for built-in tools to handle `rustdoc::` lints (see above) so
+this isn't much more of a step.
+
+How do we allow controling precedence between lints and lint groups?  We are
```

Comment:
I'm imagining these are treated as rustc flags, right? There's already a precedence between flags / code and groups

Reply:
Isn't the precedence based on the order they appear on the command line?

If so, when we translate the fields
```toml
future-incompatible = "deny"
semicolon_in_expressions_from_macros = "allow"
```
to rustc flags, how do we know it should be translated to `-D future-incompatible -A semicolon_in_expressions_from_macros`?

While to the user, the TOML appears ordered, by definition it isn't and it is dependent on whether we use `indexmap`, `BTreeMap`, or `HashMap` for what ordering we get and that would be an implementation detail bleeding out.

Reply:
> Isn't the precedence based on the order they appear on the command line?

Oh, for groups vs levels, yes

Perhaps we can specify it as groups-first-then-individuals?

Reply:
How would cargo know what is a group?  We could hard code the list and try to keep it up-to-date but that makes third-party tool lints even more difficult.

Reply:
With our [larger thread on precedence](https://github.com/rust-lang/rfcs/pull/3389#discussion_r1133265436), I'm marking this as resolved to encourage the conversation over there